metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/ViewSynchronizer.scala

```diff
@@ -0,0 +1,151 @@
+package io.iohk.metronome.hotstuff.service.sync
+
+import cats._
+import cats.implicits._
+import cats.effect.{Timer, Sync}
+import cats.data.NonEmptySeq
+import io.iohk.metronome.core.Validated
+import io.iohk.metronome.hotstuff.consensus.{Federation, ViewNumber}
+import io.iohk.metronome.hotstuff.consensus.basic.{
+  Agreement,
+  Signing,
+  QuorumCertificate,
+  Phase
+}
+import io.iohk.metronome.hotstuff.service.Status
+import io.iohk.metronome.hotstuff.service.tracing.SyncTracers
+import scala.concurrent.duration._
+import io.iohk.metronome.hotstuff.consensus.basic.ProtocolError
+
+/** The job of the `ViewSynchronizer` is to ask the other federation members
+  * what their status is and figure out a view number we should be using.
+  * This is something we must do after startup, or if we have for some reason
+  * fallen out of sync with the rest of the federation.
+  */
+class ViewSynchronizer[F[_]: Sync: Timer: Parallel, A <: Agreement: Signing](
+    federation: Federation[A#PKey],
+    getStatus: ViewSynchronizer.GetStatus[F, A],
+    retryTimeout: FiniteDuration = 5.seconds
+)(implicit tracers: SyncTracers[F, A]) {
+  import ViewSynchronizer.aggregateStatus
+
+  /** Poll the federation members for the current status until we have gathered
+    * enough to make a decision, i.e. we have a quorum.
+    *
+    * Pick the highest Quorum Certificates from the gathered responses, but be
+    * more careful with he view number as these can be disingenuous.
+    *
+    * Try again until in one round we can gather all statuses from everyone.
+    */
+  def sync: F[Status[A]] = {
+    federation.publicKeys.toVector
+      .parTraverse(getAndValidateStatus)
+      .flatMap { maybeStatuses =>
+        tracers
+          .statusPoll(federation.publicKeys -> maybeStatuses)
+          .as(maybeStatuses.flatten)
+      }
+      .map(NonEmptySeq.fromSeq)
+      .flatMap {
+        case Some(statuses) if statuses.size >= federation.quorumSize =>
+          aggregateStatus(statuses).pure[F]
+
+        case _ =>
+          // We traced all responses, so we can detect if we're in an endless loop.
+          Timer[F].sleep(retryTimeout) >> sync
+      }
+  }
+
+  private def getAndValidateStatus(
+      from: A#PKey
+  ): F[Option[Validated[Status[A]]]] =
+    getStatus(from).flatMap {
+      case None =>
+        none.pure[F]
+
+      case Some(status) =>
+        validate(from, status) match {
+          case Left((error, hint)) =>
+            tracers.invalidStatus(status, error, hint).as(none)
+          case Right(valid) =>
+            valid.some.pure[F]
+        }
+    }
+
+  private def validate(
+      from: A#PKey,
+      status: Status[A]
+  ): Either[
+    (ProtocolError.InvalidQuorumCertificate[A], ViewSynchronizer.Hint),
+    Validated[Status[A]]
+  ] =
+    for {
+      _ <- validateQC(from, status.prepareQC)(
+        checkPhase(Phase.Prepare),
+        checkSignature,
+        checkVisible(status),
+        qc =>
```

Comment:
Minor, subjective: for improved aesthetics I'd give it a name, even if used only once

Reply:
Done, thanks, it was ugly indeed. Left it under the prepare QC checks, rather than a new "status level" block, since there's just the `InvalidQuorumCertificate` error available for this; have to put the blame on one of the QCs.